[cxx-interop] Implement foreign reference types.

zoecarver opened the pull request:
This is an experimental feature to allow an attribute, `import_as_ref`, to import a C++ record as a non-reference-counted reference type in Swift.

zoecarver committed:
[cxx-interop] Implement foreign reference types.

This is an expiremental feature to allow an attribute, `import_as_ref`, to import a C++ record as a non-reference-counted reference type in Swift.

rjmccall commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4276,9 +4295,10 @@ namespace {
         if (!dc->isModuleScopeContext()) {
-          if (selfIsInOut)
+          // Don't import members of a class decl as mutating.
+          if (selfIsInOut && !isa<ClassDecl>(dc)) {
```
Maybe `selfIsInOut` should just be false when we import as a reference type.

rjmccall commented on lib/IRGen/GenCall.cpp:
```diff
@@ -3510,6 +3529,13 @@ static void externalizeArguments(IRGenFunction &IGF, const Callee &callee,
+        out.add(in.claimNext());
+        break;
+      }
+
```
We shouldn't see this.

zoecarver commented:
@rjmccall this should be ready for re-review now. I think I addressed all your comments. Let me know what you think.

Note: I still need to fix the issue with optionals (unless some macro tells us otherwise, we should be importing foreign reference types as optionals). I'll do that shortly, but I think the rest of the PR can still be reviewed without that change.

zoecarver commented on lib/SIL/Verifier/SILVerifier.cpp:
```diff
@@ -3351,9 +3353,14 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
+    // The method ought to appear in the class vtable unless it's a foreign
+    // reference type.
+    bool isForeignReferenceTy =
+        operandType.getClassOrBoundGenericClass()->getClangDecl() &&
```
But thanks for catching that!

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented on include/swift/AST/Decl.h:
```diff
@@ -3841,7 +3842,8 @@ class ClassDecl final : public NominalTypeDecl {
   bool isForeign() const {
-    return getForeignClassKind() != ForeignKind::Normal;
+    return getForeignClassKind() != ForeignKind::Normal ||
+      const_cast<ClassDecl *>(this)->isForeignReferenceType();
```
This is not ideal, but I think it's OK because we never construct a const decl, right?

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@rjmccall friendly ping :)

zoecarver commented on include/swift/AST/DiagnosticsIRGen.def:
```diff
@@ -59,5 +59,10 @@ ERROR(temporary_allocation_alignment_not_positive,none,
 ERROR(temporary_allocation_alignment_not_power_of_2,none,
       "alignment value must be a power of two", ())
 
+ERROR(foreign_reference_types_unsupported,none,
```
@rjmccall and @beccadax I need to add a test for this new error, but other than that, this should be ready for re-review.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test Linux.

zoecarver commented:
@swift-ci please smoke test Linux.

zoecarver commented:
@swift-ci please smoke test Linux.

zoecarver commented:
@swift-ci please smoke test Linux.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

hyp commented on lib/AST/Type.cpp:
```diff
@@ -5612,6 +5608,18 @@ TypeBase::getAutoDiffTangentSpace(LookupConformanceFn lookupConformance) {
+}
+
+bool CanType::isForeignReferenceType() {
+  if (auto *classDecl = getPointer()->lookThroughAllOptionalTypes()->getClassOrBoundGenericClass())
```
Looks like you duplicate the logic here. Can you call `getPointer()->isForeignReferenceType()` instead?

zoecarver commented on test/Interop/Cxx/foreign-reference/not-any-object.swift:
```diff
@@ -0,0 +1,29 @@
+public func test(_ _: AnyObject) {}
+
+// TODO: make this a better error.
+test(Empty.create()) // expected-error {{type of expression is ambiguous without more context}}
```
Add a test for casting explicitly to any object.
